keps/sig-node/2033-rootless/README.md

```diff
@@ -0,0 +1,797 @@
+
+However, the user doesn't want Kubernetes/kind/Docker/Podman to gain the root privileges
+because these components may accidentally break the host environment,
+e.g. Docker may modify the host iptables in an unexpected way and break the user's VPN connectivity.
+
+The user would want this KEP to be implemented so that he/she could run `kind` with Rootless Docker/Podman,
+which won't break the host.
+
+### Notes/Constraints/Caveats
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+TL;DR: Things that work with Rootless Docker 20.10 and Rootless Podman 2.1 will work with Rootless Kubernetes as well. Other things will not.
+
+cgroup:
+- No support for cgroup v1.
+- [Hugepages](../20190129-hugepages.md) cannot be supported because systemd doesn't support delegation of the hugetlb controller: https://github.com/systemd/systemd/issues/16325
+- Device controller cannot be supported as well, but it is not a huge deal, because non-root users don't have permission to access insecure devices anyway.
+
+Network:
+- kube-proxy needs the following `KubeletConfiguration` to avoid hitting errors during setting `sysctl` values:
+```yaml
+conntrack:
+# Skip setting sysctl value "net.netfilter.nf_conntrack_max"
+  maxPerCore: 0
+# Skip setting "net.netfilter.nf_conntrack_tcp_timeout_established"
+  tcpEstablishedTimeout: 0s
+# Skip setting "net.netfilter.nf_conntrack_tcp_timeout_close"
+  tcpCloseWaitTimeout: 0s
+```
+- Some CNI plugins might not work. Flannel (VXLAN) is known to work.
+- Limited network performance due to the slirp4netns overhead.
+  **Mitigation:** Install [`lxc-user-nic` (SETUID binary)](https://github.com/rootless-containers/rootlesskit/blob/v0.14.0-beta.0/docs/network.md#--netlxc-user-nic-experimental).
+- NodePort less than 1024 cannot be exposed. This is not a problem with the default `--service-node-port-range` configuration (30000-32767).
+  **Mitigation:** set `CAP_NET_BIND_SERVICE` file capability on `rootlesskit` binary.
+
+Volumes:
+- Block device volumes and (kernel-mode) NFS does not work, because user namespace only supports `tmpfs`, `bind`, and FUSE filesystems.
+  `emptyDir`, `hostPath`, `local`, and API volumes (`configMap`, `secret`, `downwardAPI`, ...) are known to work without any issue.
+  FUSE-based CSI volumes can be supported as well.
+  **Mitigation:** Implement CSI plugins for FUSE implementation of NFS. Or implement a "proxy" CSI plugin that talks to a privileged CSI plugin daemon.
+
+SecurityContext:
+- A container with `securityContext.privileged` cannot gain the real root privileges, obviously.
+- `runAsUser`: supported, but the number of the UID is limited by `/etc/subuid`.
+- `sysctls`: some sysctl parameters are supported, but some would fail in `EPERM`.
```

Comment:
Do we have a list of sysctls? (I think overall we would need a comprehensive document that captures what works and what doesn't for rootless.)

Reply:
Kernel should provide the list, but I don't think we have the list.
Basically, sysctl that do not result in container breakout are allowed, and others are disallowed.

Reply:
@AkihiroSuda ACK, we could use the alpha->beta to figure out what kernels things fail for which sysctl params.

Reply:
RHEL has been officially supporting Rootless Podman since RHEL 8.1, and it has exactly same restriction as this KEP, but I don't see the list of supported sysctls in RHEL documentation 

https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/building_running_and_managing_containers/starting-with-containers_building-running-and-managing-containers#special_considerations_for_rootless

Reply:
I guess the list of userns-scoped sysctls can be tracked in https://rootlesscontaine.rs/ (or eventually in kernel documentation).
Opened issue https://github.com/rootless-containers/rootlesscontaine.rs/issues/34 .